### PR TITLE
Improvements in the name and map generators

### PR DESCRIPTION
**Pull request by spixi**
This commit contains several improvements in the name generator and map generator:
- factory class to simplify creation of name generators and removing dependencies from concrete name generator classes
- the `is_valid()` method has been replaced by an exception, because that is more meaningful than using a shared pointer to delete the object, if it is not valid
- the context-free grammar based name generator now supports variables
- the map generator now uses the new features of the cfg name generator
- the redundant wml tag `[village_naming]` has been removed
- village and terrain generating rules have been ported from`/data/english.cfg` to `/data/core/macros/names.cfg`

### CODE REVIEW

File: src/generators/default_map_generator_job.cpp [hunk 1]
```diff
@@ -785,17 +762,26 @@ std::string default_map_generator_job::default_generate_map(size_t width, size_t
...
-			name_generator.reset();
-		}
-	}
-	config::attribute_value markov_list = naming.get_old_attribute("names", "male_names",
-		"[naming]male_names is deprecated, use names instead");
-	if(!markov_list.blank()) {
-		name_generator.reset(new markov_generator(utils::split(markov_list), naming["markov_chain_size"], 12));
+	name_generator_factory* generator_factory = nullptr;
+	name_generator	*base_name_generator      = nullptr,
+					*river_name_generator     = nullptr,
```
Comment (CelticMinstrel) [inline src/generators/default_map_generator_job.cpp]: Personally, I would prefer if you repeat the type name on each line here, rather than declaring several pointers in a single statement. I'm not sure if Wesnoth has an official stance on this though.

File: src/generators/default_map_generator_job.cpp [hunk 2]
```diff
@@ -1089,8 +1068,8 @@ std::string default_map_generator_job::default_generate_map(size_t width, size_t
 		// Search a path out for the road
 		pathfind::plain_route rt = pathfind::a_star_search(src, dst, 10000.0, &calc, width, height);
 
-		std::string road_base_name;
-		const std::string& name = generate_name(name_generator, "road_name", &road_base_name);
+		std::string road_base_name = base_name_generator->generate();
+		const std::string& name = road_name_generator->generate(std::map<std::string,std::string>{{{"base", road_base_name}}});
```
Comment (CelticMinstrel) [inline src/generators/default_map_generator_job.cpp]: Does this work without the cast? If not, it may be nicer to use a typedef. (I think there already is one for this particular type somewhere - `string_map`.)

File: src/utils/context_free_grammar_generator.cpp [hunk 3]
```diff
@@ -125,36 +116,66 @@ context_free_grammar_generator::context_free_grammar_generator(const std::map<st
...
-	std::string result;
-	std::map<std::string, nonterminal>::const_iterator found = nonterminals_.find(name);
-	if (found == nonterminals_.end()) {
-		lg::wml_error() << "[context_free_grammar_generator] Warning: needed nonterminal " << name << " not defined";
-		return "!" + name;
+std::string context_free_grammar_generator::print_nonterminal(const std::string& name, const std::map<std::string,std::string>* variables, uint32_t* seed, short seed_pos) const {
+	if (name.length() >= 2 && name[0] == '$') {
+		std::string variable = name.substr(1);
+		if (variables == nullptr) {
+			goto warn_missing_variable;
```
Comment (CelticMinstrel) [inline src/utils/context_free_grammar_generator.cpp]: It looks like you could eliminate these two goto statements by simply inverting your conditionals.

File: src/utils/name_generator.hpp [hunk 4]
```diff
@@ -16,20 +16,33 @@
...
+};
+
 
 class name_generator {
 public:
+	virtual std::string generate(const std::map<std::string,std::string>& variables) const = 0;
 	virtual std::string generate() const = 0;
-	virtual bool is_valid() const {return true;}
-	virtual ~name_generator() {}
+	name_generator() throw (name_generator_invalid_exception) {};
```
Comment (CelticMinstrel) [inline src/utils/name_generator.hpp]: There's no need for exception-specifications, and in fact it may be better to avoid them since I believe C++ is phasing them out in later revisions.

**Comment by CelticMinstrel** [comment]
Okay, so there's a lot of stuff here. I would have been happier if this had been split into several commits, honestly, but it may be too late to split it up now. (If you want to try, feel free, but keep in mind that splitting commits is kind of difficult in git.)

Some of the changes you've made are a little dubious as well. That's not to say that they're bad, strictly speaking, but...

1. Though exceptions are nice, I don't think they're categorically better here. The old method works just as well, right? Overuse of exceptions is also problematic when debugging, though I don't think this particular use of them would be (and Wesnoth abuses exceptions far worse in other places).
2. I don't quite understand how the name generator factory class works, but the existence of `dummy_generator` feels like a red flag. Why do you need a dummy generator? Surely there's a way to do that without creating a useless class that does nothing? Furthermore, even if you absolutely do need `dummy_generator`, I don't think there's any point in giving it a whole file-pair to itself; it could just as easily live in `name_generator_factory.cpp` (yes, a source-only class with no header).
3. When I was working on the name generators awhile ago, I was under the impression that `[village_naming]` was used in a different manner than `[naming]`. Are you quite sure it's redundant? Can you explain your reasoning in detail? Furthermore, even if it's redundant, you can't just up and remove it. There's a deprecation procedure that needs to be followed.
4. I did scan over the code, but could you explain how the CFG variables work in simple terms? If this is accepted it would need to be documented in the wiki anyway, and I'm currently on the fence about the new feature, so a clear explanation may help me better formulate an opinion on it.

All things considered, it's my opinion that this cannot be merged in its present form. The

**Comment by spixi** [comment]
@CelticMinstrel 
Thank you for your feedback.

I readded the `[village_naming]` tag again, removed the exception specifications and the `goto` statements and the typename in front of the initializer lists. I cannot remove `name_generator_factory.hpp`, because it is required to initialize a static private attribute. The `name_generator::generate` method has a parameter where you can provide a string map of variables which should be replaced. The syntax of variables is like nonterminals, but variables start with a Dollar sign, e. g. the context free grammar `main={$var} Town|City of ($var}|{$var2} {$var}` with the variables `var="New" ` and `var2="York"` would produce `York Town`, `City of York` or `New York`.

**Comment by gfgtdf** [comment]
I don'T see where the pointer returned by determine_name_generator_from_config is deleted also i wonder why base_generator_factory  is a pointer and just just a normal name_generator_factory object..

**Comment by CelticMinstrel** [comment]
> I cannot remove `name_generator_factory.hpp`, because it is required to initialize a static private attribute. 

I think you misunderstood my second point. The suggestion was not to remove _name_generator_factory.hpp_ – rather, I suggested that the contents of `dummy_name_generator.hpp` and `dummy_name_generator.cpp` should be merged into `name_generator_factory.cpp`. (Possibly defining all the functions inline, since there's little reason to define them out of line if the class and function definitions are in the same file.)

File: src/utils/dummy_generator.cpp [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+   the Free Software Foundation; either version 2 of the License, or
+   (at your option) any later version.
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY.
+   See the COPYING file for more details.
+*/
+
+/**
+ *  @file
+ *  Fallback for name generation (to guarantee, that name_generator_factory always returns a valid object reference)
```
Comment (CelticMinstrel) [inline src/utils/dummy_generator.cpp]: This leads to the following question: Why does `name_generator_factory` always need to return a valid object reference? It could easily return a pointer or a `boost::optional` instead, for example.

File: src/generators/default_map_generator_job.cpp [hunk 6]
```diff
@@ -785,17 +762,36 @@ std::string default_map_generator_job::default_generate_map(size_t width, size_t
...
+	name_generator*	bridge_name_generator          = nullptr;
+	name_generator*	mountain_name_generator        = nullptr;
+	name_generator*	forest_name_generator          = nullptr;
+	name_generator*	swamp_name_generator           = nullptr;				
+
+	if(misc_labels != nullptr) {
+		base_generator_factory = new name_generator_factory(naming, {"male", "base", "bridge", "road", "river", "forest", "lake", "mountain", "swamp"} );
+		
+		naming.get_old_attribute("base_names", "male_names", "[naming]male_names= is deprecated, use base_names= instead");
+		naming.get_old_attribute("base_name_generator", "male_name_generator", "[naming]male_names= is deprecated, use base_names= instead");
```
Comment (CelticMinstrel) [inline src/generators/default_map_generator_job.cpp]: Incorrect message. You don't really need this one though - since `male_name_generator` is new and not present in any release, it's fine to remove it outright with no deprecation message.

File: src/utils/context_free_grammar_generator.cpp [hunk 7]
```diff
@@ -30,8 +30,7 @@ context_free_grammar_generator::~context_free_grammar_generator()
 {
 }
 
-context_free_grammar_generator::context_free_grammar_generator(const std::string& source) :
-	initialized_(false)
+context_free_grammar_generator::context_free_grammar_generator(const std::string& source) throw ()
```
Comment (CelticMinstrel) [inline src/utils/context_free_grammar_generator.cpp]: This is wrong! You only use `throw()` when you want to guarantee that the function will never throw an exception, but this function is clearly throwing exceptions.

**Comment by spixi** [comment]
I made some changes:
- removed singleton and pointers 		
- removed wrong `throw()`		
- removed deprecation message

@CelticMinstrel  It is a common practise that factories produce Null objects, when they fail to produce any other object. This makes it less error prone, because you always get a valid object and you don't need to check whether it is a null pointer. It also provides a default implementation to reduce code duplication.

**Comment by spixi** [comment]
How can I fix the

`undefined reference to \`name_generator_factory::name_generator_factory(config const&, std::vector<std::string, std::allocator<std::string> >)'`

`undefined reference to \`name_generator_factory::get_name_generator(std::string)'`

in the Travis build?

The protoypes `name_generator_factory::get_name_generator(const std::string)` and `name_generator_factory::name_generator_factory(config const&, std::vector<std::string>` are inlcuded in `src/utils/name_generator_factory.hpp`

**Comment by CelticMinstrel** [comment]
> It also provides a default implementation to reduce code duplication.

Uhh... what are you talking about? The rest of what you said kinda makes sense (even if I'm not sure I agree with it), but I can see no way in which `dummy_name_generator` provides a default implementation or reduces code duplication (in fact, it kinda increases code duplication). Anyway, if you wanted a default implementation for something, you would put it in the `name_generator` base class.

File: src/scripting/game_lua_kernel.cpp [hunk 8]
```diff
@@ -4374,62 +4375,62 @@ static int intf_name_generator(lua_State *L)
...
+
 	static const char*const generic_err = "error initializing name generator";
 	if(!gen) {
 		lua_pushstring(L, generic_err);
 		return lua_error(L);
 	}
-	// We set the metatable now, even if the generator is invalid, so that it
-	// will be properly collected if it was invalid.
-	luaL_getmetatable(L, "name generator");
-	lua_setmetatable(L, -2);
```
Comment (CelticMinstrel) [inline src/scripting/game_lua_kernel.cpp]: Uhhh... why did you remove these two lines? They're crucial for this to work. (The comment should be retained too.)

File: src/scripting/game_lua_kernel.cpp [hunk 9]
```diff
@@ -4374,62 +4375,62 @@ static int intf_name_generator(lua_State *L)
...
 		}
-	} else {
-		return luaL_argerror(L, 1, "should be either 'markov_chain' or 'context_free'");
 	}
+	catch (const name_generator_invalid_exception& ex) {
+		lua_pushstring(L, ex.what());
+		return lua_error(L);		
+	}
+
 	static const char*const generic_err = "error initializing name generator";
```
Comment (CelticMinstrel) [inline src/scripting/game_lua_kernel.cpp]: If I understand the situation correctly, this conditional below is now unreachable, and this error message will never be seen. Is that accurate?

**Comment by CelticMinstrel** [comment]
I think perhaps an error should be raised in the CFG generator if you attempt to add a nonterminal beginning with $ - a simple conditional throw statement in the condition block at line 41 would suffice, I think.

**Comment by CelticMinstrel** [comment]
Regarding the Travis issue, it's probably because you didn't add the new file to _src/SConscript_. You should also add it to _src/CMakeLists.txt_.

**Comment by spixi** [comment]
I added the `name_generator_factory.cpp` file into `src/SConscript` (It was already in `src/CMakeLists.txt`). Apparently travis uses SCons, but I used CMake for my local build.

`$` sign in nonterminal is now checked.

The `dummy_generator` class has been removed, the base class has now a default implementation.

Furthermore I removed the unused branch in `game_lua_kernel.cpp` and added the `lua_get/setmetatable` again.

I should really consider to install SCons, but it requires Python 2.7 and its kitchen sink. (At the moment I have only Python 3.4 installed.)

The coding in `context_free_grammar_generator.cpp` really needs a code review. I kept it as-is, because I don't know Wesnoth's development guidelines and I don't wanted to break it.

File: src/utils/context_free_grammar_generator.cpp [hunk 10]
```diff
@@ -125,36 +124,60 @@ context_free_grammar_generator::context_free_grammar_generator(const std::map<st
...
+
+		lg::wml_error() << "[context_free_grammar_generator] Warning: needed variable" << variable << " not defined";
+		return "!" + variable;
 	}
-	const context_free_grammar_generator::nonterminal& got = found->second;
-	unsigned int picked = seed[seed_pos++] % got.possibilities_.size();
-	if (seed_pos >= seed_size) seed_pos = 0;
-	if (picked == got.last_) {
-		picked = seed[seed_pos++] % got.possibilities_.size();
+	else {
```
Comment (CelticMinstrel) [inline src/utils/context_free_grammar_generator.cpp]: ...you know, this else isn't actually needed since the if block always returns. It would make the diff clearer, as well.

**Comment by CelticMinstrel** [comment]
Well, I don't see any problems left, though I'd prefer someone more familiar with map generation to review that part of the PR.

**Comment by gfgtdf** [comment]
Hmm what random number is used to generate the names dorign map generation? Usually the map generator has its own random generator default_map_generator_job::rng_ that could be used but note that just like with unit the number of random calls should be independent on the name generator (which might depend on trasnlations) to make random mape rproducable with a given seed.

Other than that, i doubt there there is someone really familiar with the map generator code, i recently refactoeed it a bit, but i mainly moved functions around, not changing the map generators logic. It would be nice if random map generators coudl be tested before this is merged, that is, the mainline randommp maps with both scenario_generation and map_generation.

**Comment by CelticMinstrel** [comment]
Even if there isn't anyone familiar with the map generator, it would be nice to know that someone other than me has looked through this in some detail (as opposed to just glancing through it quickly). If you've done that, then great.

**Comment by sigurdfdragon** [comment]
While I'm not qualified to review the c++ code, I could take a look and test that it doesn't obviously break anything with the mainline random maps or my own random maps add-on, but that's about it.

**Comment by gfgtdf** [comment]
No i didnd't look into this in deati. It'd be nice if you'd test it in mainline and umc campaigns

**Comment by CelticMinstrel** [comment]
@sigurdfdragon That would be great! Perhaps not quite as great as someone scanning the code for obvious errors, but I suppose it may have to do. It'd also be great if someone tested campaign usages of the random map generator, for example in HTTT.

**Comment by gfgtdf** [comment]
httt only used teh cave map generator which doen't use the code in default_map_generator_job.cpp

**Comment by sigurdfdragon** [comment]
@CelticMinstrel I've got an add-on where I can test campaign usage too.

**Comment by spixi** [comment]
I just noted that there is a small change in the interface which means that old add-ons may not be compatible with this PR.

Because I moved the village and terrain names from `english.cfg` to the `[naming]` and `[village_naming]` tags it is possible to have different naming schemes per map generator. As long the add-on used the macro `{VILLAGE_NAMES}`, everything is okay. But If it uses the old interface to define custom names and only defines `male_names=` (which is now an alias for `base_names=`), but not `bridge_names`, `swamp_names`, `village_names`, `village_forest_names` and so on (or their corresponding `*_name*generator` equivalent, they won't get any names at all.

**Comment by CelticMinstrel** [comment]
Hmm. I think it's actually quite likely that an add-on would use `male_names=` directly instead of the `{VILLAGE_NAMES}` macro - this is arguable the point of having the `[naming]` tag in the first place, really. So, that could be a problem.

I'm not sure what's the best solution, but here's one idea - add a `[naming]` child to the game config (in _data/game_config.cfg_) which includes `{VILLAGE_NAMES}`. Then that can be used as a fallback if the `[naming]` child in the generator does not include some key. I'm not sure if the `{VILLAGE_NAMES}` macro is visible in that file though, it should be easy to make it so if it's not already.

(If anyone has any better ideas, feel free to state them.)

**Comment by sigurdfdragon** [comment]
I tested this rebased against current master and with the following random maps (default generator): mainline maps, umc maps in mp, and umc campaign maps in mp. There are problems.

Umc campaign random maps cause a crash when loading a map. (they use the map_generation=default path.) Backtrack to be posted.

Mainline & umc random map scenario seem ok. (they use the scenario_generation=default path.)

@CelticMinstrel also, it doesn't solve https://gna.org/bugs/?24709. I still need to have [naming] tags or it will crash upon loading.

As far as the quality of the naming output, it looks ok at a glance, and that's all I can say about it. I don't use naming in any of my random map add-ons & I'm working around needing naming tags because of bug#24709 by clearing the labels in a prestart event. I added tags back in briefly to test the above.

**Comment by sigurdfdragon** [comment]
Program received signal SIGSEGV, Segmentation fault.
0x00b4f688 in default_map_generator_job::default_generate_map (this=0x283ec8,
    width=90, height=90, island_size=0, island_off_center=0, iterations=1125,
    hill_size=6, max_lakes=40, nvillages=22, castle_size=6, nplayers=2,
    roads_between_castles=true, labels=0x0, cfg=...)
    at src\generators\default_map_generator_job.cpp:1079
1079                    std::string road_name = base_name_generator->generate();

(gdb) bt
#0  0x00b4f688 in default_map_generator_job::default_generate_map (
    this=0x283ec8, width=90, height=90, island_size=0, island_off_center=0,
    iterations=1125, hill_size=6, max_lakes=40, nvillages=22, castle_size=6,
    nplayers=2, roads_between_castles=true, labels=0x0, cfg=...)
    at src\generators\default_map_generator_job.cpp:1079
#1  0x00b3d4c0 in default_map_generator::generate_map (this=0x4d56a990,
    labels=0x0, randomseed=...)
    at src\generators\default_map_generator.cpp:410
#2  0x00b3d07c in default_map_generator::create_map[abi:cxx11](boost::optional<u
nsigned int>) (this=0x4d56a990, randomseed=...)
    at src\generators\default_map_generator.cpp:351
#3  0x00aa3cb5 in random_generate_map (parms="default", cfg=...)
    at src\generators\map_create.cpp:60
#4  0x004ad8ca in saved_game::expand_random_scenario (this=0x418945f4)
    at src\saved_game.cpp:422
#5  0x00903713 in mp::initial_level_config (state=...)
    at src\game_initialization\mp_game_utils.cpp:61
#6  0x008d28b3 in ng::connect_engine::connect_engine (this=0x4d56a7c8,
    state=..., first_scenario=false, campaign_info=0x0)
    at src\game_initialization\connect_engine.cpp:96
#7  0x006d81c0 in campaign_controller::play_game (this=0x28770c)
    at src\game_initialization\playcampaign.cpp:375
#8  0x006d224a in enter_connect_mode (video=..., game_config=..., state=...,
    wesnothd_connection=0x0, local_players_only=true)
---Type <return> to continue, or q <return> to quit---
    at src\game_initialization\multiplayer.cpp:493
#9  0x006d2cfb in enter_configure_mode (video=..., game_config=...,
    state=..., wesnothd_connection=0x0, local_players_only=true)
    at src\game_initialization\multiplayer.cpp:591
#10 0x006d27b6 in enter_create_mode (video=..., game_config=..., state=...,
    wesnothd_connection=0x0, local_players_only=true)
    at src\game_initialization\multiplayer.cpp:547
#11 0x006d3d11 in mp::start_local_game (video=..., game_config=..., state=...)
    at src\game_initialization\multiplayer.cpp:730
#12 0x004445a1 in game_launcher::play_multiplayer (this=0x41894598)
    at src\game_launcher.cpp:880
#13 0x00404ed4 in do_gameloop (
    args=std::vector of length 3, capacity 4 = {...}) at src\wesnoth.cpp:800
#14 0x00406894 in SDL_main (argc=argc@entry=3, argv=argv@entry=0x3f00c8)
    at src\wesnoth.cpp:1071
#15 0x00bc961f in main_utf8 (argv=0x3f00c8, argc=<optimized out>)
    at ../src/main/windows/SDL_windows_main.c:126
#16 WinMain@16 (hInst=0x400000, hPrev=0x0,
    szCmdLine=0x41bc34fe "--config-dir C:\\Wesnoth\\Master\\userdata", sw=10)
    at ../src/main/windows/SDL_windows_main.c:189
#17 0x012b57bd in main ()
(gdb) cont
Continuing.
[Thread 7468.0x1e9c exited with code 0]
[Thread 7468.0x19c8 exited with code 3221225477]
[Thread 7468.0x464 exited with code 3221225477]
[Thread 7468.0x1e18 exited with code 3221225477]
[Thread 7468.0x1138 exited with code 3221225477]
[Thread 7468.0x510 exited with code 3221225477]
[Thread 7468.0x1cbc exited with code 3221225477]
[Thread 7468.0x1e2c exited with code 3221225477]
[Thread 7468.0xe2c exited with code 3221225477]
[Inferior 1 (process 7468) exited with code 030000000005]

Problem signature:
  Problem Event Name:	APPCRASH
  Application Name:	wesnoth-debug.exe
  Application Version:	1.13.5.0
  Application Timestamp:	5757406c
  Fault Module Name:	wesnoth-debug.exe
  Fault Module Version:	1.13.5.0
  Fault Module Timestamp:	5757406c
  Exception Code:	c0000005
  Exception Offset:	0074f688
  OS Version:	6.1.7601.2.1.0.256.1
  Locale ID:	1033
  Additional Information 1:	ac88
  Additional Information 2:	ac8897866058893e0fcc5ef9b0ddf66d
  Additional Information 3:	e7bf
  Additional Information 4:	e7bfa9cd8cb3502dd3be89fde0a41d4e

**Comment by gfgtdf** [comment]
this needs to be fixed before this can be merged.

**Comment by spixi** [comment]
I'll try to have a look on this on Saturday. I have no time atm.

**Comment by spixi** [comment]
I see the issue. The problem is, that there are missing keys in `[naming]` tag. I tried to add a default config to `data/_main.cfg`, but don't know how to access this from the default name generator job (which only knows the children of the `[generator]` tag.

The following keys are mandantory at the moment, I try to make them optional soon.

`[naming]`
- `base_names` or `base_name_generator` or `male_names` or `male_name_generator`
- `bridge_name_generator` or `bridge_names`
- `road_name_generator` or `road_names`
- `river_name_generator` or `river_names`
- `forest_name_generator` or `forest_names`
- `lake_name_generator` or `lake_names`
- `mountain_name_generator` or `mountain_names`
- `swamp_name_generator` or `swamp_names`


`[village_naming]`
- `village_name_generator` or `village_name_names`
- `village_lake_name_generator` or `village_lake_names`
- `village_river_name_generator` or `village_river_names`
- `village_bridge_name_generator` or `village_bridge_names`
- `village_grassland_name_generator` or `village_grassland_names`
- `village_forest_name_generator` or `village_forest_names`
- `village_hill_name_generator` or `village_hill_names`
- `village_mountain_name_generator` or `village_mountain_names`
- `village_mountain_anonymous_name_generator` or `village_mountain_anonymous_names`
- `village_road_name_generator` or `village_road_names`
- `village_swamp_name_generator` or `village_swamp_names`

Another possibility would be just generating no name, if there is no such key found (similar to races, which use a dummy name generator)

**Comment by CelticMinstrel** [comment]
You know, you could make the village_ prefix optional there.

The way I was proposing adding a default config was to put it in the `[game_config]` tag, which I think is in a file called _game_config.cfg_. If you did that, it would be somehow accessible in an object in the source code, declared in _resources.hpp_. (I think either the variable name or type name is `game_config`.)

**Comment by spixi** [comment]
`village_name_names` was a typo in my comment, sorry for that. I don't think that the prefixes are optional, because both tags use the `{VILLAGE_NAMING}` macro at the moment, which probably will be empty after setting the default values. However we still should prevent generating names instead of crashing when the default parameters are missing.

**Comment by gfgtdf** [comment]
i don't think it makes sense to provide defautl names for the map generator, if no name(generator)s are given, then teh map generator shoudl just not generate map labels for names.

**Comment by CelticMinstrel** [comment]
The issue here is compatibility - as it currently stands, any older random map generators that do use names will no longer work as the creator intended (assuming that they don't use the `{VILLAGE_NAMING}` macro, which in my opinion is a fairly good assumption), because they don't contain the extra name specifications.

It would be helpful if we could get some kind of survey of map generator uses in addons on the 1.12 addons server, as well as mainline uses. If it turns out that everyone _does_ use the `{VILLAGE_NAMING}` macro, then this might be easier to resolve acceptably.

I think that if `[naming]` and `[village_naming]` have different expected keys, then it's a little silly to use the same macro for the standard contents of both. I do think that the `village_` prefix on those keys is mainly pointless, as well.

...I also suddenly had the thought that integrating variables into the CFG generator is actually kind of redundant. You could have just as easily used the existing WML variable substitution mechanism - use the CFG generator to generate a string containing WML variable substitutions, then substitute the variables into the resulting string. Maybe it's a little late to suggest making that change, though...

**Comment by spixi** [comment]
I am sorry to tell you that I have no time at the moment. Maybe I will have a look at it during the next two weeks.

Regards,

Spixi

**Comment by spixi** [comment]
My idea is the following:

I move the default implementations of the name generators back to english.cfg into the keys `[default_naming]` and `[default_village_naming]`, so I can access them with
	config = game_config_manager::get()->game_config().child("language")


The user-defined name generators are merged with the game config ones. If neither a user-defined, nor a game config name generator is found, we generate no name.

Then I make `{{VILLAGE_NAMING}}` an empty macro for backward-compatibility reasons.

Is that okay? If yes, I will do the change in the next weeks (maybe next Saturday).

**Comment by CelticMinstrel** [comment]
That sounds okay, but I think I'd put base_names / base_name_generator in the `{{VILLAGE_NAMING}}` macro rather than in english.cfg.

I'm still having second thoughts on the redundancy of your variables implementation, too.

**Comment by Vultraz** [comment]
Please rebase this branch with master.

**Comment by gfgtdf** [comment]
you messed up the intention in some files and added some large file 'install_manifest.txt' this must be fixed before mering.

**Comment by spixi** [comment]
I somehow messed up my local repository. I am working on it.

**Comment by spixi** [comment]
running `git fsck`

**Comment by spixi** [comment]
So, I think, it seems to be okay now?

**Comment by CelticMinstrel** [comment]
Okay, I want to merge this pull request soon. However, I am against the variable support in the CFG generator. Is there any reason why you can't do something like this?

1. Rig the CFG generator to generate strings containing WML varaible substitution syntax, eg `bridge_name_generator= _ << main=$base|’s Bridge|$base|’s Crossing >>
2. Substitute the variables into the resulting string by calling `utils::interpolate_variables_into_string(cfg_output, variables)`. This is defined in _formula/string_utils.hpp_.

Besides that, my only remaining problem is, why do you need the village_ prefix on all the keys in `[village_naming]`? That they are village-related is implied by the tag name and doesn't need to be repeated, as far as I can see. I'd prefer if the prefix was dropped, unless you can provide a good reason not to (maybe I've missed something).

**Comment by spixi** [comment]
Hi @CeltricMinistrel,

I care about your remarks tomorrow. A small question: Should it be possible to support dynamic variables, e. g.

```
my_generator = <<
main=$foo{var}
var=1,2,3
>>
```

which would be identical too
```
my_generator="main=$foo1,$foo2,$foo3"
```

or should the variable substitution come first, which allows things like:

```
my_generator = <<
main="{foo$var|}"
foo1="foo1"
foo2="foo2"
>>
```

Thank you for your feedback.

**Comment by CelticMinstrel** [comment]
Okay, so I'm glad you mentioned this, because it brought forward an issue with the idea that I suggested. The issue is quite simple, really - the `|` character is important both in the name generator and in variable substitution, with different meanings. (It's actually optional in variable substitution, but there exist cases where it is required, eg `$name|marsh`.)

For your actual question, I favour the first way - that is, the CFG could include sequences that vary which construct variable names from smaller parts. Note that your example of what it would be identical to is actually incorrect, as it uses commas instead of vertical bar as the separator.

Regarding the problem of `|`, I can see two options:
1. Implement an escaping mechanism in the CFG parser to allow special characters to be specified as literals instead of with their special meaning. There are two obvious ways to implement an escaping mechanism.
2. Select a character (eg `#`) to replace the variable-substitution meaning of `|` only in generators. Then the procedure would be: generate a string from the generator; replace all `#` with `|`; and finally substitute variables.
3. We could ignore the problem and just posit that variable substitutions requiring `|` are not supported, but since there are already some cases in english.cfg that would require it, I don't consider this a good option.

The two obvious ways of implementing an escaping mechanism are:
1. Use an escape character; backslash is the most obvious choice, but percent, hash, or even at could work too. (It should not be a character that would commonly be used in natural text. This excludes ampersand, forward slash, and all sentence punctuation.) When the escape character is encountered, if it is followed by a special character (which includes itself as well as `={}|`), then the escape character is ignored and the special character is inserted into the current buffer. If it is not followed by a special character, it could either be ignored or inserted into the current buffer.
2. Use braces. This method might not allow for escaping the braces themselves (though it could allow the sequences `{{}` and `{}}` for that, but that might be confusing). The most important result of this is that the sequence `{|}` would expand to a vertical bar in the output. (You could also support `{=}`, though that would only be useful if these escapes are valid on the left hand side as `=` on the right hand side is already taken as a literal.)
3. Define an implicit nonterminal in every generator that expands to a vertical bar. With this method, there's no actual escape mechanism, but the sequence `{pipe}` (or similar) will expand to a vertical bar in the output. This might actually be the easiest solution.

Anyone have thoughts on this?

**Comment by Vultraz** [comment]
What about a double pipe (`||`) for the generator, and a single pipe (`|`) for variable substitution?

**Comment by ln-zookeeper** [comment]
If you're talking about making || have special meaning in variable substitution in general: || is very commonly used in cases like $foo_$side_number|| and those cannot be broken.

**Comment by spixi** [comment]
I totally forgot about this. The vertical bar was the main problem, that's why I used the comma instead in my comment above.

For escaping I have the following suggestion:

There are the predefined nonterminals:
```
( = "{"
) = "}"
! = "|"
```

This would change the syntax from
```
forest_name_generator= _ << main={$base} Forest|{$base}’s Forest >>
```
to

```
forest_name_generator= _ << main=$base Forest|$base{!}’s Forest >>
```

The variable substitution will be executed after the name generator.

**Comment by CelticMinstrel** [comment]
That seems okay to me. I'd prefer `{|}` to `{!}`, but I guess either is acceptable.

(Incidentally, I don't think the vertical bar is actually required when it's followed by a single quote / apostrophe - only when followed by a letter, digit, or underscore, and possibly also when followed by a dot or square bracket.)

@ln-zookeeper That's not what we were talking about, so no worries.

**Comment by spixi** [comment]
shit, I don't know what I did ...

**Comment by CelticMinstrel** [comment]
Two issues I can still see:
1. Why is there an error if `$` is used in a nonterminal? I suppose it's not really a major issue, but it does seem weird. I could let this pass if there's a good reason.
2. This still conflicts with some changes made in master. In particular, I recently moved the Lua API implementation for the name generators from _game_lua_kernel.cpp_ to _lua_kernel_base.cpp_, so if you could update the PR for that, it'd be great. I also made a few bugfixes to the function, so please make sure that you don't accidentally revert those.

**Comment by spixi** [comment]
I fixed the error with the ```$``` sign. I forgot to adjust it, when I reverted the syntax for the variables back to WML variables syntax.

@CelticMinstrel: Do you think you could do the changes with _lua_kernel_base.cpp_ and the bugfixes mentioned in you comment?

**Comment by CelticMinstrel** [comment]
I suppose I could give that a try.

**Comment by spixi** [comment]
Thank you

**Comment by CelticMinstrel** [comment]
This has now been merged.

**Comment by sigurdfdragon** [comment]
The crash I reported on June 7th above with this pr was NOT fixed. I have opened bug report: https://gna.org/bugs/index.php?24864
Please fix or revert!

**Comment by CelticMinstrel** [comment]
It seems to be fixed by 465ca498b6d6 - road names were trying to be generated even if no rules for generating names had been given.

**Comment by sigurdfdragon** [comment]
confirmed fixed, thanks
